glide.yaml

```diff
@@ -1,8 +1,11 @@
 package: github.com/percona/pmm-client
 import:
 - package: github.com/fatih/color
+  version: ^1.5.0
```

Comment:
Personally I'm against using versions. This only causes problems. What benefit this brings? Golang is not scripting language. If there is a break in API your code will fail immediately after update - then you can either update your code or lock library to old version. In other scripting languages it also helps with determining which version of shared library to use - I'm not even sure if glide scans for other libraries glide.yamls - and even if it does then you can't be sure that dependencies also use glide. You will also loose the ability to easily upgrade to never versions of libs as `glide update` will only update within semversions - there is no easy way to update glide.yaml, you will need to do it manually.

Reply:
> If there is a break in API your code will fail immediately after update

I wish… It is still possible for API to be cnahged, but for compilation to suceed. Consider, for example, interfaces and type assertions: when something stopped implementing some interfce, and your code works differently. Or consider varargs. Or reflection.

> You will also loose the ability to easily upgrade to never versions of libs as glide update will only update within semversions - there is no easy way to update glide.yaml, you will need to do it manually.

That's why I love dep more and mode. :)

The biggest benefit of using tagged version is that most people put more care about creating tags (and thinking in the process about docs, tests, compatiblity, etc) than to default branch.
Another benefit: it's easier for us. Updating from 1.3 to 1.4 is much easier than from some randomly-looking commit to another one. And if upstream cares about SemVer and keeps CHANGELOG… ❤️

Reply:
I know you can have incompatibilities even if compilation succeeds... but it's extremely rare. I'm afraid that introducing so strict versions will lead to not updating libs at all. I don't see a reason for introducing this - this was never a problem, so why we try to solve something that doesn't need solving by introducing more work? Why try to solve problems we don't have?

Reply:
That `^` means `>= 1.5.0, < 2.0.0`, so it's not really strict.
And we do have this problem: `prometheus/client_golang` 0.8 is not compatible with `master`, the same for some gRPC-related packages.

We want our users to use specific versions, not `master`, right? The same logic applies there: it is much easier to report errors against releases, not some random commit between them.

Reply:
Srsly Aleksi, I know what `^` means. If you need to lock some version, then lock that one package, not every package.

> We want our users to use specific versions, not master, right?
> The same logic applies there: it is much easier to report errors against releases, not some random commit between them.

You are mixing things. End user is not dev, system package manager aka deb/rpm is not golang package manager. Dev dependencies required for compilation are not runtime dependencies on linux systems. If you had *one* issue with incompatible lib then lock that lib to older version (yes, it's fine with range constraint), but don't lock everything, it completely doesn't make sense especially that many golang libs follow go-way and promise always compatibility in their master branches. If you have compatibility issue with one lib, lock it, at least it will be visible in glide.yaml that there is never version but we didn't update our code base to never version. 

Srsly, I don't know why we are loosing time on something that doesn't need fixing, but do as you wish - pmm-client works perfectly fine without this.

Reply:
> Srsly Aleksi, I know what ^ means.

I'm just saying that it's not a "strict version" as you put it.

> but don't lock everything

I didn't lock a version. I want to update from released version to released version, not from some development commit to some other commit.

> that many golang libs follow go-way and promise always compatibility in their master branches

Do you have some data to back this up? My experience tells me the opposite. From the top of my head: https://github.com/prometheus/client_golang/compare/v0.8.0...master There are a lot of breaking changes, and there will be more before 0.9.

> Srsly, I don't know why we are loosing time on something that doesn't need fixing, but do as you wish - pmm-client works perfectly fine without this.

Ok. May you please do me a favor? Can you add github.com/percona/pmm-managed to `vendor/` without updating other deps?

Reply:
I can understand your annoyance. I hope you will agree with my arguments later. Probably, one way to better understand me is to actually try to add a new dependency without updating others. Glide just can't do that if versions are not specified.